File: src/main/java/rak/planetGenerator/generators/TemperatureMapGenerator.java
package rak.planetGenerator.generators;

import rak.utility.MathFunctions;

public class TemperatureMapGenerator {
	private int temperature = 50;
	private int temperatureVariance = 50;
	
	private int minTemperature;
	private int maxTemperature;
	
	private int[][] temperatureMap;
	
	private static final float TEMPERATURE_FACTOR = 1.4f;
	
	public int[][] generateTemperatureMap(int[][] heightMap) {
		initializeTemperatureMap(heightMap);
		setMinAndMaxTemperature();
		populateTemperatures(heightMap);
		
		return temperatureMap;
	}

	private void initializeTemperatureMap(int[][] heightMap) {
		int density = heightMap.length;
		temperatureMap = new int[density][density];
	}

	private void setMinAndMaxTemperature(){
		minTemperature = temperature - temperatureVariance;
		maxTemperature = temperature + temperatureVariance;
	}
	
	private void populateTemperatures(int[][] heightMap) {
		for (int y=0; y<temperatureMap.length; y++){
			int rowTemperature = createRowTemperature(y);
			
			populateColumnTemperatures(heightMap, y, rowTemperature);
		}
	}

	private int createRowTemperature(int y) {
		float latitude = MathFunctions.getPercent(y, temperatureMap.length);
		int rowTemperature = maxTemperature; 
		rowTemperature -= TEMPERATURE_FACTOR * Math.abs(temperatureVariance * latitude);
		return rowTemperature;
	}
	
	private void populateColumnTemperatures(int[][] heightMap, int y, int rowTemperature) {
		for (int x=0; x<temperatureMap.length; x++){
			int altitude = heightMap[x][y];
			int localTemperature = getLocalTemperature(rowTemperature, altitude);
			
			temperatureMap[x][y] = localTemperature;
		}
	}
	
	private int getLocalTemperature(int columnTemperature, int altitude) {
		int localTemperature = columnTemperature;
		localTemperature -= getAltitudeVariance(altitude);
		localTemperature = clampToPossibleTemperature(localTemperature);
		
		return localTemperature;
	}

	private int getAltitudeVariance(int altitude) {
		if (altitude > 0){
			return temperatureVariance * altitude/100;
		}
		return 0;
	}
	
	private int clampToPossibleTemperature(int localTemperature) {
		return MathFunctions.clamp(localTemperature, minTemperature, maxTemperature);
	}
	
	public void setTemperature(int temperature){
		this.temperature = temperature;
	}
	
	public void setTemperatureVariance(int temperatureVariance){
		this.temperatureVariance = temperatureVariance;
	}
}

File: src/main/java/rak/planetGenerator/generators/TemperatureMapGenerator.kt
package rak.planetGenerator.generators

import rak.utility.MathFunctions

class TemperatureMapGenerator {
    private var temperature = 50
    private var temperatureVariance = 50

    private var minTemperature: Int = 0
    private var maxTemperature: Int = 0

    private var temperatureMap: Array<IntArray>? = null

    fun generateTemperatureMap(heightMap: Array<IntArray>): Array<IntArray>? {
        initializeTemperatureMap(heightMap)
        setMinAndMaxTemperature()
        populateTemperatures(heightMap)

        return temperatureMap
    }

    private fun initializeTemperatureMap(heightMap: Array<IntArray>) {
        val density = heightMap.size
        temperatureMap = Array(density) { IntArray(density) }
    }

    private fun setMinAndMaxTemperature() {
        minTemperature = temperature - temperatureVariance
        maxTemperature = temperature + temperatureVariance
    }

    private fun populateTemperatures(heightMap: Array<IntArray>) {
        for (y in temperatureMap!!.indices) {
            val rowTemperature = createRowTemperature(y)

            populateColumnTemperatures(heightMap, y, rowTemperature)
        }
    }

    private fun createRowTemperature(y: Int): Int {
        val latitude = MathFunctions.getPercent(y, temperatureMap!!.size)
        var rowTemperature = maxTemperature
        rowTemperature -= (TEMPERATURE_FACTOR * Math.abs(temperatureVariance * latitude)).toInt()
        return rowTemperature
    }

    private fun populateColumnTemperatures(heightMap: Array<IntArray>, y: Int, rowTemperature: Int) {
        for (x in temperatureMap!!.indices) {
            val altitude = heightMap[x][y]
            val localTemperature = getLocalTemperature(rowTemperature, altitude)

            temperatureMap!![x][y] = localTemperature
        }
    }

    private fun getLocalTemperature(columnTemperature: Int, altitude: Int): Int {
        var localTemperature = columnTemperature
        localTemperature -= getAltitudeVariance(altitude)
        localTemperature = clampToPossibleTemperature(localTemperature)

        return localTemperature
    }

    private fun getAltitudeVariance(altitude: Int): Int {
        return if (altitude > 0) {
            temperatureVariance * altitude / 100
        } else 0
    }

    private fun clampToPossibleTemperature(localTemperature: Int): Int {
        return MathFunctions.clamp(localTemperature, minTemperature, maxTemperature)
    }

    fun setTemperature(temperature: Int) {
        this.temperature = temperature
    }

    fun setTemperatureVariance(temperatureVariance: Int) {
        this.temperatureVariance = temperatureVariance
    }

    companion object {

        private val TEMPERATURE_FACTOR = 1.4f
    }
}
